### PR TITLE
Fairly extensive rewrite

### PR DESCRIPTION
**Pull request by Ghostbird**
The main reason why I started this rewrite was that I needed a newer Xamarin OSS to be able to build AAB files. The old one doesn't seem to know the `AndroidPackageFormats` option.

Then I ran into several other issues for example all recent Xamarin OSS build artefacts are unsigned. However they also include a deb file which greatly simplifies the installation.
I ran into dependency issues and I changed the base image to Debian instead of Ubuntu. Ultimately that didn't solve anything, but 
I didn't feel like changing it back. I did have to use `libzip4` instead of `libzip5` as the latter isn't packaged with Debian. The dependency issues were resolved by installing OpenJDK 11 SDK. This turned out to satisfy the Xamarin OSS deb file's dependencies.

I was a bit confused by the convoluted repository structure, and after reading some best-practices on the Docker website, I rewrote the entire structure to a single multi-stage Dockerfile.

While working on this fork I encountered a lot of fork-unfriendly hard-coded values. I extracted all of those into variables and created a `default.sh` file that gives them default values, unless they're already set. For example from a sourced `.env` file.

Since I was reworking the scripts anyway, I put the api level and build-tools definitions in an associative array, and iterate over its keys in numerical order in the `build.sh` and `push.sh` scripts. To add support for a new API level, only a single key-value needs to be added to the `defaults.sh` definition of `BUILD_TOOLS`.

### CODE REVIEW

**Comment by tafilz** [comment]
Hey, thanks for the PR!

I was new to docker and publishing to docker hub and this was needed for a student project that used gitlab ci to build the app. The project structure was … a bit complicated. I agree on that :)

These changes look good to me and learned something new. I will test things on my end as soon as i am home.

File: defaults.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+# Note: The BUILD_TOOLS list is the definition. The apiLevels variable is a numerically sorted list of its keys.
+# Bash associative arrays are in hash order. By extracting the apiLevels into a normal array,
+# we can iterate over them in numerically ascending order.
+API_LEVELS=()
+for apiLevel in $(tr ' ' '\n' <<< ${!BUILD_TOOLS[@]} | sort -n)
+do
+  API_LEVELS+=( "${apiLevel}" )
+done
+API_LEVELS
```
Comment (tafilz) [inline defaults.sh:51]:
```suggestion
```
I think this was some kind of left over?

Comment (Ghostbird) [reply defaults.sh]: Indeed, I've removed it.

**Comment by Ghostbird** [comment]
Sure. I'll do it tomorrow.

**Comment by Ghostbird** [comment]
Date tagging and arbitrary tagging has been added. It can be controlled with these variables:

- EXTRA_TAGS: an array of tag values
- TAG_WITH_DATE: set to `1` or `yes` to automatically add YYYYMMDD formatted date stamp as tag
  You can pass `--tag-with-date` to the `build.sh` and `push.sh` which does the same as setting TAG_WITH_DATE

**Comment by Ghostbird** [comment]
There's no more recent build of Xamarin OSS available, yet I wish there was. This one suffers from https://github.com/xamarin/Xamarin.Forms/issues/15352. Which isn't a deal-breaker, but it means that exceptions thrown in the debugger are obscured, and you need to jump to the correct context in the stacktrace and evaluate `this.SourceException` to get to the real exception.

**Comment by tafilz** [comment]
> Can we request a build pipeline run + retain somewhere?

Good question. I still didn't find out which of these builds are the "current" release.

**Comment by Ghostbird** [comment]
I think I'll have to rewrite it again. The good news is. It's become even simpler than I thought. The reason why the OSS Xamarin Android image on the build page is out of date, is that they've changed the release mechanism. According to https://github.com/xamarin/xamarin-android/issues/7235 we can simply run:

```bash
# dotnet workload install android
```
To install Xamarin Android on a Linux machine.

I'll experiment with it once I have some time.

**Comment by Ghostbird** [comment]
Ok, an annoying thing is that for Android API level 32 and 33 we need to install dotnet workloads `android-32` and `android-33` respectively.
Also using this mechanism would be a breaking change for users of these docker images, since you'll need to make a change like this:
```patch
-    <TargetFrameworkVersion>v12.0</TargetFrameworkVersion>
+    <TargetFramework>net6.0-android</TargetFramework>
```
What do you think we should do. I think we want to support the new functionality. I'm not sure how to organise it so that it doesn't break stuff for existing users.

**Comment by tafilz** [comment]
> Also using this mechanism would be a breaking change for users of these docker images, since you'll need to make changes to the project set-up.

I didn't touch the newest xamarin yet. What changes are needed? The old images are going to stay with the date tag. 

> Should we add -net6 to all tags for the new mechanism?
> Or should we go for a new repository: xamarin-android-net6 entirely?

I think we should add those changes here. It's easier to maintain the Xamarin Android OSS build without digging into the build pipelines to find a working version. 

What do you think? Should we add these breaking changes here and add a notice for existing users in the dockerhub readme and here, or add an extra build with the `-net6` tag (maybe instead of new repo and addtional branch?)

**Comment by Ghostbird** [comment]
Well, my conclusion is that it's nigh impossible to get the new mechanism to work. So let's stay at the current mechanism for now, and try to tackle the newer mechanism at some other time.

**Comment by tafilz** [comment]
I'll push the retagged old images later tonight and merge and build and push the new images, if i don't find any issues.

**Comment by Ghostbird** [comment]
Have a look at the line that says:
```docker
ENV PATH=/xamarin/bin/Release/bin:$PATH
```
I retained it from the original, but I think it's no longer needed with the `.deb` install. `/xamarin` doesn't seem to exist.

**Comment by tafilz** [comment]
Yeah. Can be safely removed

**Comment by tafilz** [comment]
Images are getting pushed right now

**Pull request merged**